File: common/src/main/java/com/ultreon/mods/advanceddebug/api/common/Multiplier.java
package com.ultreon.mods.advanceddebug.api.common;

import com.ultreon.mods.advanceddebug.api.client.formatter.IFormatterContext;
import com.ultreon.mods.advanceddebug.api.util.MathUtil;

import java.util.Objects;

public record Multiplier(double value) implements IFormattable {
    @Override
    public void format(IFormatterContext ctx) {
        if (MathHelper.getDecimalPlaces(this.value) == 0) {
            ctx.number(Math.round(this.value)).operator("×");
        } else {
            ctx.number(MathUtil.roundTo(this.value, 5)).operator("×");
        }
    }

    public Percentage percentage() {
        return new Percentage(this.value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || this.getClass() != o.getClass()) return false;
        Multiplier that = (Multiplier) o;
        return Double.compare(that.value, this.value) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.value);
    }
}

File: common/src/main/kotlin/com/ultreon/mods/advanceddebug/api/common/Multiplier.kt
package com.ultreon.mods.advanceddebug.api.common

import com.ultreon.mods.advanceddebug.api.client.formatter.IFormatterContext
import com.ultreon.mods.advanceddebug.api.util.MathUtil
import java.util.*
import kotlin.math.roundToLong

class Multiplier(value: Double) : IFormattable {
    val value: Double
    val percentage: Percentage get() = (value * 100).percent

    override fun format(ctx: IFormatterContext) {
        if (MathHelper.getDecimalPlaces(value) == 0) {
            ctx.number(value.roundToLong()).operator("×")
        } else {
            ctx.number(MathUtil.roundTo(value, 5)).operator("×")
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as Multiplier
        return that.value == value
    }

    override fun hashCode(): Int {
        return Objects.hash(value)
    }

    init {
        this.value = value
    }
}

val Number.multiplier
    get() = Multiplier(this.toDouble())
